stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,654 @@
+    passed from Driver to Driver without excessive calls to KMS.encrypt/decrypt, this "data" in
+    the Credentials can be either encrypted or not. It is up to the code that constructs the
+    Credentials object to know which.
+
+    When retrieving the raw, unencrypted data from a Credentials object, use the following code:
+
+        if credentials.is_encrypted():
+            return credentials.get_data_kms_decrypted()
+        else:
+            return credentials.data()
+    """
+
+    def __init__(self, data, is_encrypted=False, region=None):
+        """
+        Args:
+            data (object): A json serializable object, or a string.
+            is_encrypted (bool): Pass True if the input data is encrypted with KMS. False otherwise.
+            region (str): AWS Region. Only required if is_encrypted=True.
+        """
+        self._data = data
+        self._is_encrypted = is_encrypted
+        self._region = region if is_encrypted else None  # No use for region if unencrypted
+
+    def is_encrypted(self):
+        """True if this Credentials object is encrypted. False otherwise.
+
+        Returns:
+            bool
+        """
+        return self._is_encrypted
+
+    def data(self):
+        """
+        Returns:
+            str: The raw text data of this Credentials object, encrypted or not. This may be
+                unusable if encrypted, but can be passed to another Driver for storage.
+        """
+        return self._data
+
+    def get_data_kms_decrypted(self):
+        """Returns the data of this Credentials objects, decrypted with KMS.
+
+        This does not mutate the internals of this Credentials for safety. It simply returns
+        the decrypted payload. It is up to the called to safely manage the payload.
+
+        Returns:
+            str|None: The decrypted payload of this Credentials object, if it is encrypted. If it is
+                not encrypted, then will return None and log an error.
+        """
+        if not self._is_encrypted:
+            LOGGER.error('Cannot decrypt Credentials as they are already decrypted')
+            return None
+
+        try:
+            client = boto3.client('kms', config=default_config(region=self._region))
+            response = client.decrypt(CiphertextBlob=self._data)
+            return response['Plaintext']
+        except ClientError as err:
+            LOGGER.exception('an error occurred during credentials decryption: %s', err.response)
+            return None
+
+    def encrypt(self, region, kms_key_alias):
+        """Encrypts the current Credentials.
+
+        Calling this method will entirely change the internals of this Credentials object.
+        Subsequent calls to .data() will return encrypted data.
+        """
+        self._region = region
+        if self.is_encrypted():
+            return
+
+        self._is_encrypted = True
+        if not self._data:
+            return
+
+        creds_json = json.dumps(self._data)
```

Comment:
we should use 'minimal' json here when dumping (`json.dumps(self._data, separators=(',', ':'))`) to reduce file size as much as possible. this isn't really a problem with s3, but could be with SSM due to sorta low limits. the gain won't be very huge anyway so maybe not a huge deal